crates/vault-registry/src/types.rs

```diff
@@ -102,19 +104,242 @@ pub type DefaultVaultId<T> = VaultId<<T as frame_system::Config>::AccountId, Cur
 
 pub type DefaultVaultCurrencyPair<T> = VaultCurrencyPair<CurrencyId<T>>;
 
+pub mod v2 {
+    use super::*;
+
+    #[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, Default, TypeInfo)]
+    struct WalletV2 {
+        // store all addresses for `report_vault_theft` checks
+        pub addresses: BTreeSet<BtcAddress>,
+        // we use this public key to generate new addresses
+        pub public_key: BtcPublicKey,
+    }
+
+    #[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo)]
+    #[cfg_attr(feature = "std", derive(Debug))]
+    struct VaultV2<AccountId, BlockNumber, Balance, CurrencyId: Copy> {
+        /// Account identifier of the Vault
+        pub id: VaultId<AccountId, CurrencyId>,
+        /// Bitcoin address of this Vault (P2PKH, P2SH, P2WPKH, P2WSH)
+        pub wallet: WalletV2,
+        /// Current status of the vault
+        pub status: super::VaultStatus,
+        /// Block height until which this Vault is banned from being used for
+        /// Issue, Redeem (except during automatic liquidation) and Replace.
+        pub banned_until: Option<BlockNumber>,
+        /// Number of tokens pending issue
+        pub to_be_issued_tokens: Balance,
+        /// Number of issued tokens
+        pub issued_tokens: Balance,
+        /// Number of tokens pending redeem
+        pub to_be_redeemed_tokens: Balance,
+        /// Number of tokens that have been requested for a replace through
+        /// `request_replace`, but that have not been accepted yet by a new_vault.
+        pub to_be_replaced_tokens: Balance,
+        /// Amount of collateral that is available as griefing collateral to vaults accepting
+        /// a replace request. It is to be payed out if the old_vault fails to call execute_replace.
+        pub replace_collateral: Balance,
+        /// Amount of collateral locked for accepted replace requests.
+        pub active_replace_collateral: Balance,
+        /// Amount of collateral that is locked for remaining to_be_redeemed
+        /// tokens upon liquidation.
+        pub liquidated_collateral: Balance,
+    }
+
+    type DefaultVaultV2<T> = VaultV2<
+        <T as frame_system::Config>::AccountId,
+        <T as frame_system::Config>::BlockNumber,
+        BalanceOf<T>,
+        CurrencyId<T>,
+    >;
+
+    pub fn migrate_v2_to_v3<T: Config>() -> frame_support::weights::Weight {
```

Comment:
I like this migration pattern, did you use a particular reference for this?

Reply:
Not sure what pattern you mean but I based the migration function itself on https://github.com/paritytech/substrate/blob/master/frame/scheduler/src/lib.rs#L582 . 

Personally I also like the way I organized all the migration code into a single module, that'll make removal later on super easy